### PR TITLE
Fix one case of extra calls to CCurlFile::GetMimeType

### PR DESCRIPTION
**Pull request by Paxxi**
Actually use the mime type that is passed in.
Don't blindly refresh mime type unless it's missing

@stefansaraev @wsnipex  can you verify?

### CODE REVIEW

**Comment by stefansaraev** [comment]
:+1 I confirm

**Comment by fritsch** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
@fritsch to be clear, this is not the grooveshark issue but a separate one causing the same behaviour. I think i have found the groove shark issue as well but havent but together a patch yet

**Comment by fritsch** [comment]
@Paxxi It makes sense anyways or not?

**Comment by Paxxi** [comment]
Yeah this is going in, just thought I'd mention it since I saw you active in grooveshark ticket

File: xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 1]
```diff
@@ -105,7 +107,10 @@ CDVDInputStream* CDVDFactoryInputStream::CreateInputStream(IDVDPlayer* pPlayer,
   {
     if (item.IsType(".m3u8"))
       return new CDVDInputStreamFFmpeg();
-    item.FillInMimeType();
+
```
Comment (wsnipex) [inline xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp]: cosmetics, there is a tab here.
and thanks for looking into this :)

File: xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 2]
```diff
@@ -45,6 +45,8 @@ CDVDInputStream* CDVDFactoryInputStream::CreateInputStream(IDVDPlayer* pPlayer,
 {
   CFileItem item(file.c_str(), false);
 
+  item.SetMimeType(content);
```
Comment (FernetMenta) [inline xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp]: this is a sensitive region. i only have a tablet and can't have a closer look. please wait with merge

Comment (ace20022) [inline xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp]: I can't see/find a place where the mime type is used before line 108. Why is this call needed? Maybe I've overlooked something?

**Comment by FernetMenta** [comment]
what exactly should this fix?
I am quite sure it would break MMS streams because of bypassing the logic we have in FillInMimeType

**Comment by stefansaraev** [comment]
> what exactly should this fix?

this.
http://mirrors.kodi.tv/docs/python-docs/14.x-helix/xbmcgui.html#ListItem-setMimeType

an api methot, that was broken for quite long.

**Comment by FernetMenta** [comment]
imo this api method should go away. it does more harm than any good.

**Comment by wsnipex** [comment]
no, its needed for some use cases. Grooveshark is one example: Its API provides a "one time key" thats only valid for a single http request. If the mimetype is not set, we'll do a HEAD request in CurlFile::Stat() to detect it, which will burn/invalidate the key.

**Comment by FernetMenta** [comment]
ok, but then it should be implemented properly, not as a hack. if there is a requirement not to stat, the reason should be obvious and reflected by the api as such.

**Comment by wsnipex** [comment]
whats wrong with simply not doing Stat when the mimetype is already set?

**Comment by FernetMenta** [comment]
there is a reason why we have this behavior in GetMimeType for MMS. As said already, bypassing this may break other behavior. 
Further there are other scenarios where we need to stat. Hiding the original intent is always a bad idea and will break sooner or later. What you want to accomplish here is avoiding stat. This is not the same as bypassing existing calls.

File: xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp (hunk 2 above)
Comment (Paxxi) [inline xbmc/cores/dvdplayer/DVDInputStreams/DVDFactoryInputStream.cpp]: It's not needed until later but it felt reasonable to set when we construct our fileitem

**Comment by Paxxi** [comment]
Isnt mms handled earlier with the check for mms:// ? Why does this method take a mimetype if we dont use it?

**Comment by FernetMenta** [comment]
did you read the comments in CFileItem::FillInMimeType? this is non trivial behavior and should not exposed via the api where it gets used without this knowledge.
and again, your actual problem seems be be something else.

**Comment by Paxxi** [comment]
I've traced through the whole call stack and everything is correct up until this method, the correct mimetype is passed in but the method ignores it. I'm open to different solutions if this will cause issues but ignoring the mimetype if we have it seems like strange behaviour, I would expect that we honor it. If there's known issues with some urls I think that should be handled elsewhere in the stack so that the mimetype is correct once we get to this function.

**Comment by Paxxi** [comment]
I would also like to point out that using the mimetype in a method designed to take a mimetype isn't what I'd like to call a hack, it seems like the designed for experience, not using it and blindly calling FillMimeType is a hack, possibly much needed but just my two cents :)

**Comment by FernetMenta** [comment]
again: passing the mime type is not appropriate for all cases because external components do not know the behavior we need. assume some addon passes video/x-ms*. this pr would break behavior. options won't be stripped either as current code does.

**Comment by wsnipex** [comment]
thats the addons authors problem. If he sets a wrong mimetype, his addon will fail.

**Comment by FernetMenta** [comment]
no, it is not. we don't sacrifice behavior for others. we also pass codec information as a hint to player but this does not mean that we don't query for actual values. same for mime type. look at it as a hint. the implementation can stat an url as often as it wants if there is not explicit rule not to do so.

**Comment by wsnipex** [comment]
then suggest a "proper" fix for the issue. Just saying its wrong to let addons override the mimetype is not going to work. Fact is that adding the unconditional Stat is a regression.

**Comment by Paxxi** [comment]
No there's no rule but our current behaviour is incompatible with some server setups, @stefansaraev bumped into this issue but says the server admins have resolved it on their end. We have another case of this issue that breaks the grooveshark plugin because their server setups are weird and or broken.

I do believe that if an addon passes in a broken mimetype we should fail, however that should be introduced in a new function. I don't think we should change the rules and break existing addons on this.

@FernetMenta would you be ok with adding something like TrustMyMimeType and exposing that to addons? then we can let them set it and we let them break if they do a mistake and we can keep our current behvaiour unchanged.

**Comment by FernetMenta** [comment]
I ask again: what is the actual problem. @wsnipex told me that it is NOT mime type but an additional call to stat. @wsnipex please read what I wrote. I did suggest a proper fix. If you want to avoid this call to stat, implement this and not something else.

**Comment by Paxxi** [comment]
The issue is that some services use onetime access keys but they're not smart enough to differentiate between get and head. When we stat to get the mimetype we silently invalidates the access key and the request fails. Since the stat is caused by us ignoring the set mimetype and querying the server for it the proper solution to avoid the stat seemed to be to actually use the mimetype and that's what this pr does

**Comment by FernetMenta** [comment]
Ok, so we are given an URL we only can use for streaming. Access of this URL for any other purpose is not permitted. This is a different story than avoiding request of mime type. Someone may implement some other feature which accesses this URL -> this would break Grooveshark again.
It is much better to explicitly set this information "this is streaming url only".

**Comment by stefansaraev** [comment]
most so-called "streaming urls" are perfectly fine with "stat call" (HEAD requests sent from curl). some are not. that's why there is ("was", because its broken after 708db324) setMimeType() python api method exposed. the method is documented. for years. it's up to addon devs to decide whether they want to use the method, or not. 

the method was added in Gotham, and broken in Helix.

if mimetype is not set from outside (addon), there is no change in behaviour.

ping @tamland

**Comment by fritsch** [comment]
Btw. an example of how to do something like this "explicitely" without breaking given API given on another topic here: https://github.com/xbmc/xbmc/pull/6759

**Comment by tamland** [comment]
I haven't followed this entire discussion, but this fix seems good too me. I intended that `FillInMimeType` line to only do so when all else fail. In addition to what's mentioned here, another use for the setMimeType api is to explicitly *avoid* the expensive stat call altogether. It even says so in the docs.

**Comment by FernetMenta** [comment]
> but this fix seems good too me

and again: it is not for the reasons given

**Comment by stefansaraev** [comment]
@FernetMenta sorry, but I cant understand why you dont want this in.

actualy. this is a fix for a regression, introduced in a PR (pull/4729) that fixed something else, that "fix" should br wrong by design then and should be reverted and re-done properly ?

**Comment by FernetMenta** [comment]
this pr would break other things. pr 4729 broke something only worked by chance. it will happen again if not fixed properly. fix the root cause and don't go for quick hacks. the actual issue is that some urls must not be touched by anything other than streaming and the api lacks flagging this info.

**Comment by stefansaraev** [comment]
ok. then @tamland should re-think the whole thing and prepare a more acceptable fix for the api bug.

**Comment by fritsch** [comment]
We broke existing API for Helix - no matter if it that method was "good or bad" before, that should not happen and 14.2 is our last chance to get that fixed for a whole lot addons using it - without this there is no chance to set content externally.

The mentioned PR is just reverting old behaviour and should be "helix only" in my eyes. 

That said I am with @FernetMenta that we need an API change to explicitely mark such streams, which should then go to v15.

**Comment by FernetMenta** [comment]
yes. note that there are other methods like GetContent which do also access http header.

**Comment by FernetMenta** [comment]
"yes" was referring to @stefansaraev comment.

**Comment by fritsch** [comment]
Obviously clear :-) You don't like to touch half baked stuff that need a correct fix in my proposed way ... I wanted to come to a solution to concentrate on the real fix ...

**Comment by Paxxi** [comment]
We can probably agree that this fix won't go in so the next question is, how much control would we want to give to addons? I can think of two things that makes sense in the current context
ONE_TIME_URL = don't touch the url for any reason because it'll be invalid afterwards.
TRUST_MY_MIME_TYPE = trust what I've set and don't try and figure it out

**Comment by stefansaraev** [comment]
@Paxxi if you ask me, ONE_TIME_URL doesnt make sense, if kodi trusts the mimetype. again, we have a method to "force" kodi to trust the mimetype, how it's called is another story.

**Comment by FernetMenta** [comment]
@stefansaraev mimetype is not the primary issue here. there are other reasons why player requests http header apart from mime type.

**Comment by FernetMenta** [comment]
@Paxxi I think something like ONE_TIME_URL is ok. what is the requirement for TRUST_MY_MIME_TYPE? for the case mentioned here option 1 should be enough

**Comment by Paxxi** [comment]
I dont have a case for it currently but im thinking it can be useful for misconfigured servers that send wrong mimetype, with this the addon can force one to possibly work around future issues. Exposing it would be a support burden going forward so we should be sure we want to support it beforehand. I'm Just throwing my ideas out there currently to see if anyone has oppinions for or against :)

**Comment by FernetMenta** [comment]
Being prepared for such a requirement is good but unless we have an actual requirement we should not implement it. It would lead to dead code.

**Comment by fritsch** [comment]
Concerning the bug I talked above: http://trac.kodi.tv/ticket/15702#comment:19 via: 
```
if (!content.empty()) {
   item.SetMimeType(content);
} else {
   item.FillInMimeType();
}
```

**Comment by FernetMenta** [comment]
but you don't know if this breaks other files, do you? for the record: I don't hold you back submitting this to Helix branch. it is up to RM whether to take this risk or not.

**Comment by fritsch** [comment]
No obviously I don't know. For now we only know 100% sure - that we broke public API in our Helix release, everything else is speculating only.

**Comment by FernetMenta** [comment]
I don't see API broken. 

> If known prehand, this can avoid xbmc doing HEAD requests to http servers to figure out file type

"it can avoid" does not mean that it avoids it in any case.

**Comment by fritsch** [comment]
Yeah replace the "can avoid" to "will never avoid and in fact is a noop" then we really did not break anything ...

**Comment by FernetMenta** [comment]
the facts are clear:
1)
the API did never promise that it won't request header, hence those addons only worked by chance. 
2)
you provide a "fix" and do not know whether it breaks other behavior

what criteria are those which rank Grooveshark over other behavior?

**Comment by fritsch** [comment]
Yes - fully right, good summary, thx for that. I will buy a picture frame, print it and pin it into my living room.

**Comment by stefansaraev** [comment]
@FernetMenta of course, setMimeType apidocs say "may" not "should". but please dont forget we also have setProperty() method.

**Comment by denhams** [comment]
The documentation says "this can avoid xbmc doing HEAD requests". That was the case in Gotham. In Helix I don't think it is the case - it _never_ avoids HEAD requests. I think they will always be made by the player. For me that's a broken API.

**Comment by FernetMenta** [comment]
> For me that's a broken API.

nobody asked for your opinion

**Comment by Paxxi** [comment]
@FernetMenta behave, he's just trying to get his addon working
@ghost this fix will almost certainly not go in, we're trying to figure out how to solve it without breaking anything else and currently we're leaning towards extending the python api with a new method.

**Comment by Montellese** [comment]
I'm not claiming to have any knowledge about the DVDPlayer internals but from an API PoV having a method to specifically set a property and then ignore it is either bad API design or a bad/bugged implementation (either in the API implementation or in whatever is called by the API implementation).

IMO the API should behave in a way that calling `setMimeType()` will basically enforce that MIME type (that doesn't mean that the player can't do a HEAD request due to other reasons). If an addon uses that method and provides a bad MIME type it's misusing the API and is broken. How the MIME type is passed on to the the player is a different question.

So basically there are (at least) two issues here
* `setMimeType()` should enforce the passed in MIME type
* the API needs a way to disable HEAD request

**Comment by FernetMenta** [comment]
Have you ever seen a big label "this is my mime type" when playing a video? :)
I think none of you have. DVDPlayer is free to use this property in any way it wants in order to play a given file. If the best strategy is to ignore the passed value, this is fine. The API doc states it is a hint.

**Comment by fritsch** [comment]
And dvdplayer said: Let there be mime type and it saw that it was good ...

@FernetMenta: Problem here is: The file _does_not_play_ cause the player
tries to know it better which it does not ... so I second what @Montellese
said

2015-03-23 8:53 GMT+01:00 Rainer Hochecker <notifications@github.com>:

> Have you ever seen a big label "this is my mime type" when playing a
> video? :)
> I think none of you have. DVDPlayer is free to use this property in any
> way it wants in order to play a given file. If the best strategy is to
> ignore the passed value, this is fine. The API doc states it is a hint.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6753#issuecomment-84867568>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by FernetMenta** [comment]
> so I second what @Montellese said

I know and that's what I am saying all the time. mimetype is not the issue here.
